### PR TITLE
Wayland Support

### PR DESCRIPTION
**Pull request by jwinarske**
To build on Linux pass `-DMBGL_WITH_EGL=ON`

-expects GLFW built for Wayland
-defaults to Wayland over X11

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -9,6 +9,8 @@ option(MBGL_WITH_OPENGL "Build with OpenGL renderer" ON)
 option(MBGL_WITH_EGL "Build with EGL renderer" OFF)
 option(MBGL_WITH_OSMESA "Build with OSMesa (Software) renderer" OFF)
 option(MBGL_WITH_WERROR "Make all compilation warnings errors" ON)
+option(MBGL_WITH_X11_SUPPORT "Build with X11 Support" OFF)
```
Comment (ntadej) [inline CMakeLists.txt:12]: I wonder if it should be just `MBGL_WITH_X11` like other options?

Also should these be only in `linux.cmake` as they do not seem to be relevant to other platforms?

File: platform/linux/linux.cmake [hunk 2]
```diff
@@ -4,11 +4,21 @@ find_package(ICU OPTIONAL_COMPONENTS uc)
...
-find_package(X11 REQUIRED)
+if (MBGL_WITH_X11_SUPPORT)
+    find_package(X11 REQUIRED)
+endif ()
 find_package(Threads REQUIRED)
 
 pkg_search_module(LIBUV libuv REQUIRED)
 
+if (MBGL_WITH_WAYLAND_SUPPORT)
+    add_compile_definitions(
```
Comment (ntadej) [inline platform/linux/linux.cmake:15]: We try to avoid global definitions. Are they really needed or can they be added to the relevant target (maybe even `mbgl-compile-options` would do)?

**Review by ntadej — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 1 above)
Comment (jwinarske) [reply CMakeLists.txt:12]: good point. will move

File: platform/linux/linux.cmake (hunk 2 above)
Comment (jwinarske) [reply platform/linux/linux.cmake:15]: I was avoiding having to define them twice.  Switched to target_compile_definitions for each

Comment (ntadej) [reply platform/linux/linux.cmake:15]: If you define them as `PUBLIC` they will get propagated with dependencies.

Comment (jwinarske) [reply platform/linux/linux.cmake]: yes I'm very familiar with CMake

**Comment by jwinarske** [comment]
Pushed changes

File: platform/linux/linux.cmake (hunk 2 above)
Comment (ntadej) [reply platform/linux/linux.cmake]: Sorry, I just wanted to be helpful 😊

**Review by ntadej — APPROVED** [review]
Thank you for the contribution! This is OK from my side now.

**Comment by ntadej** [comment]
Actually I see no CI pipeline spawned so I'll tag @acalcutt and @louwers for feedback before merging.

**Comment by jwinarske** [comment]
Build passed
https://github.com/maplibre/maplibre-gl-native/actions/runs/4707278115/jobs/8349053136

Let me know if a Linux CI job is desired

**Comment by ntadej** [comment]
I'm more worried that this may affect Node.js as Wayland is the default now (but maybe it does not matter there).

**Review by louwers — CHANGES_REQUESTED** [review]
We use `xvfb-run` to run the render tests on Linux (not on CI right now).

So I suggest we keep X11 as the default for now.

**Comment by louwers** [comment]
I cannot push to your branch. Could you add `"platform/linux/**" on these two places so the `node-ci` workflow runs?

https://github.com/maplibre/maplibre-gl-native/blob/main/.github/workflows/node-ci.yml#L10

https://github.com/maplibre/maplibre-gl-native/blob/main/.github/workflows/node-ci.yml#L34

**Review by ovivoda — COMMENTED** [review]
Looks legit! Let's check if CI/CD passes then we can merge.

**Comment by jwinarske** [comment]
@louwers pushed

**Review by ovivoda — APPROVED** [review]

File: platform/glfw/glfw_view.cpp [hunk 3]
```diff
@@ -559,9 +559,9 @@ void GLFWView::updateFreeCameraDemo() {
 }
 
 mbgl::Color GLFWView::makeRandomColor() const {
-    const auto r = static_cast<float>(std::rand()) / RAND_MAX;
-    const auto g = static_cast<float>(std::rand()) / RAND_MAX;
-    const auto b = static_cast<float>(std::rand()) / RAND_MAX;
+    const auto r = static_cast<float>(std::rand()) / static_cast <float> (RAND_MAX);
```
Comment (louwers) [inline platform/glfw/glfw_view.cpp:562]:
```suggestion
    const auto r = static_cast<float>(std::rand()) / static_cast<float>(RAND_MAX);
```

File: platform/glfw/glfw_view.cpp [hunk 4]
```diff
@@ -559,9 +559,9 @@ void GLFWView::updateFreeCameraDemo() {
...
+    const auto g = static_cast<float>(std::rand()) / static_cast <float> (RAND_MAX);
```
Comment (louwers) [inline platform/glfw/glfw_view.cpp:563]:
```suggestion
    const auto g = static_cast<float>(std::rand()) / static_cast<float>(RAND_MAX);
```

File: platform/glfw/glfw_view.cpp [hunk 5]
```diff
@@ -559,9 +559,9 @@ void GLFWView::updateFreeCameraDemo() {
...
+    const auto b = static_cast<float>(std::rand()) / static_cast <float> (RAND_MAX);
```
Comment (louwers) [inline platform/glfw/glfw_view.cpp:564]:
```suggestion
    const auto b = static_cast<float>(std::rand()) / static_cast<float>(RAND_MAX);
```

**Review by louwers — APPROVED** [review]

**Comment by louwers** [comment]
@jwinarske Thanks for contributing.
